---
title: "FigureS9_forebrain_midbrain_late"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This documents details the steps for the generation of pseudotemporal expression profiles described in  "Temporal patterning of the central nervous system by a shared transcription factor code" by Sagner et al. 2021.

```{r Declare libraries and functions, message=FALSE, warning=FALSE, include=FALSE}
## Declare libraries ----

dir <- "/Users/j76630as/Documents/tTF_paper_2020/scRNAseq/"

setwd(paste0(dir, "/output/"))

library(Biobase)
library(slingshot)
library(Seurat)
library(dplyr)
library(scales)
library(loomR)
library(ggplot2)
library(tibble)
library(viridis)
library(pbapply)
library(ggrepel)

## Declare functions ----

#' Assign a color to each cell based on some value
#' 
#' @param cell_vars Vector indicating the value of a variable associated with cells.
#' @param pal_fun Palette function that returns a vector of hex colors, whose
#' argument is the length of such a vector.
#' @param ... Extra arguments for pal_fun.
#' @return A vector of hex colors with one entry for each cell.

cell_pal <- function(cell_vars, pal_fun,...) {
  if (is.numeric(cell_vars)) {
    pal <- pal_fun(100, ...)
    return(pal[cut(cell_vars, breaks = 100)])
  } else {
    categories <- sort(unique(cell_vars))
    pal <- setNames(pal_fun(length(categories), ...), categories)
    return(pal[cell_vars])
  }
}

## function for plotting genes in pseudotime from spinal cord scRNAseq data (Delile et al.; Development, 2019)
plot.gene.in.pseudotime <- function(gene = c('Ascl1','Nfib'),
                                    pt.mtx = pt,
                                    seurat.object = V2.neurons,
                                    exclude.timepoint = c(9.5),
                                    exclude.curve = 'curve4',
                                    plot.title = 'dI5') {
  if(is.null(exclude.curve)) {
    print("No curve excluded!")
  } else {
    print(paste0('Exclude ', exclude.curve))
    pt.mtx <- pt.mtx[, -which(colnames(pt.mtx) %in% exclude.curve)]
  }
  
  if(length(gene >1)){
    data <- t(GetAssayData(object = seurat.object, assay = 'RNA')[gene, ])
  } else {
    data <- as.data.frame(GetAssayData(object = seurat.object, assay = 'RNA')[gene, ])
    colnames(data) <- gene
  }
  
  pt.mtx <- do.call(cbind, list(pt.mtx, data,
                                timepoint = seurat.object$timepoint)) %>%
    as.data.frame() %>%
    tidyr::gather(key = curve, pseudotime, -gene, -timepoint) %>%
    tidyr::gather(key = gene_name, gene_expression, -timepoint, -curve, -pseudotime)
  
  if(is.null(exclude.timepoint)) {
    print("No timepoints excluded!")
  } else {
    print(paste0("Exclude timepoint ", exclude.timepoint))
    pt.mtx <- pt.mtx[-which(pt.mtx$timepoint %in% exclude.timepoint),]
  }
  
  pt.mtx$timepoint <- factor(as.character(pt.mtx$timepoint), levels = unique(pt.mtx$timepoint))
  pt.mtx$gene_name <- factor(pt.mtx$gene_name, levels = gene)
  
  pt.mtx <- pt.mtx[!is.na(pt.mtx$pseudotime),]
  
  gg <- ggplot(pt.mtx, aes(x = pseudotime, y = gene_expression, color = timepoint)) +
    geom_smooth(method = 'loess') +
    facet_wrap(~ gene_name)
  
  df <- as.data.frame(ggplot_build(gg)[[1]])
  
  df$gene <- factor(gene[df$PANEL], levels = unique(gene[df$PANEL]))
  
  range <- df %>%
    mutate(bin = cut(x, 30)) %>%
    group_by(gene, group, bin) %>%
    summarise(mean = mean(y)) %>%
    ungroup() %>%
    group_by(gene) %>%
    summarise(min = min(mean),
              max = max(mean))
  
  df$min <- unlist(lapply(df$gene, function(x) {
    return(range[range$gene == x, "min"])
  }))  
  
  df$max <- unlist(lapply(df$gene, function(x) {
    return(range[range$gene == x, "max"])
  }))  
  
  df$norm <- (df$y-df$min) / (df$max - df$min) 
  
  df$timepoint <- unique(pt.mtx$timepoint)[df$group]
  
  ggplot(df, aes(x = cut(x, 30), y = gene, fill = norm)) +
    geom_tile(expand = FALSE) +
    scale_fill_viridis() +
    scale_y_discrete(expand = c(0,0)) +
    theme_bw() +
    labs(title = plot.title) +
    xlab('Pseudotime per timepoint') +
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          axis.title.y = element_blank(),
          panel.grid.major=element_blank(),
          panel.grid.minor=element_blank(),
          panel.spacing.x = unit(0, "cm"),
          panel.spacing.y = unit(0, "cm"),
          plot.margin = unit(c(0, 0, 0, 0), "cm"),
          panel.border = element_rect(colour = "black", fill=NA, size=1),
          plot.background = element_blank()) +
    facet_wrap(~ timepoint, nrow = 1)
} 

### plot pseudotime of neuronal populations from loom files
### works as plot.gene.in.pseudotime
plot.gene.in.pseudotime.loom <- function(gene = c('Ascl1','Nfib'),
                                         pt.mtx = pt,
                                         seurat.object = V2.neurons,
                                         exclude.timepoint = c("e9.0"),
                                         exclude.curve = NULL,
                                         plot.title = 'Cajal-Retzius neurons',
                                         order.timepoints = c('e10.0', 'e11.0', 'e12.0', 'e12.5', 'e13.0', 'e13.5')) {
  if(is.null(exclude.curve)) {
    print("No curve excluded!")
  } else {
    print(paste0('Exclude ', exclude.curve))
    pt.mtx <- pt.mtx[, -which(colnames(pt.mtx) %in% exclude.curve)]
  }
  
  if(length(gene >1)){
    data <- t(GetAssayData(object = seurat.object, assay = 'RNA', slot = 'scale.data')[gene, ])
  } else {
    data <- as.data.frame(GetAssayData(object = seurat.object, assay = 'RNA')[gene, ])
    colnames(data) <- gene
  }
  
  pt.mtx <- do.call(cbind, list(data.frame(pt.mtx), data,
                                timepoint = seurat.object$age)) %>%
    as.data.frame() %>%
    tidyr::gather(key = curve, pseudotime, -gene, -timepoint) %>%
    tidyr::gather(key = gene_name, gene_expression, -timepoint, -curve, -pseudotime)
  
  if(is.null(exclude.timepoint)) {
    print("No timepoints excluded!")
  } else {
    print(paste0("Exclude timepoint ", exclude.timepoint))
    pt.mtx <- pt.mtx[-which(pt.mtx$timepoint %in% exclude.timepoint),]
  }
  
  pt.mtx <- pt.mtx %>%
    dplyr::mutate(age = gsub("e", "", timepoint)) %>%
    dplyr::mutate(age = as.numeric(age))
  
  pt.mtx <- pt.mtx[order(pt.mtx$age),]
  
  pt.mtx$timepoint <- factor(as.character(pt.mtx$timepoint), levels = unique(pt.mtx$timepoint))
  pt.mtx$gene_name <- factor(pt.mtx$gene_name, levels = gene)
  
  pt.mtx <- pt.mtx[!is.na(pt.mtx$pseudotime),]
  
  gg <- ggplot(pt.mtx, aes(x = pseudotime, y = as.numeric(gene_expression), color = timepoint)) +
    geom_smooth(method = 'loess') +
    facet_wrap(~ gene_name)
  
  df <- as.data.frame(ggplot_build(gg)[[1]])
  
  df$gene <- factor(gene[df$PANEL], levels = unique(gene[df$PANEL]))
  
  range <- df %>%
    mutate(bin = cut(x, 30)) %>%
    group_by(gene, group, bin) %>%
    summarise(mean = mean(y)) %>%
    ungroup() %>%
    group_by(gene) %>%
    summarise(min = min(mean),
              max = max(mean))
  
  df$min <- unlist(lapply(df$gene, function(x) {
    return(range[range$gene == x, "min"])
  }))  
  
  df$max <- unlist(lapply(df$gene, function(x) {
    return(range[range$gene == x, "max"])
  }))  
  
  df$norm <- (df$y-df$min) / (df$max - df$min) 
  
  df$timepoint <- unique(pt.mtx$timepoint)[df$group]
  
  ggplot(df, aes(x = cut(x, 30), y = gene, fill = norm)) +
    geom_tile(expand = FALSE) +
    scale_fill_viridis() +
    scale_y_discrete(expand = c(0,0)) +
    theme_bw() +
    labs(title = plot.title) +
    xlab('Pseudotime per timepoint') +
    theme(axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          axis.title.y = element_blank(),
          panel.grid.major=element_blank(),
          panel.grid.minor=element_blank(),
          panel.spacing.x = unit(0, "cm"),
          panel.spacing.y = unit(0, "cm"),
          plot.margin = unit(c(0, 0, 0, 0), "cm"),
          panel.border = element_rect(colour = "black", fill=NA, size=1),
          plot.background = element_blank()) +
    facet_wrap(~ timepoint, nrow = 1)
} 

### function to plot pseudotime curves

plot.pseudotime.curves <- function(number.columns = 3,
                                   slingshot.object = sds,
                                   cell.colors = cell_colors_clust,
                                   pal = viridis(100, end = 0.95)) {
  
  pt.mtx <- slingPseudotime(slingshot.object)
  nms <- colnames(pt.mtx)
  number.rows <- ceiling((length(nms) + 1)/number.columns) ### we add 1 so that we can plot in addition the curves on top of the clustering
  
  par(mfrow = c(number.rows, number.columns))
  
  for (i in 1:length(nms)) {
    colors <- pal[cut(pt.mtx[,i], breaks = 100)]
    plot(reducedDim(slingshot.object), col = colors, pch = 16, cex = 0.5, main = nms[i])
    lines(sds, linInd = i)
  }
  plot(reducedDim(slingshot.object), col = cell_colors_clust, pch = 16, cex = 0.5, main = "Seurat clusters")
  lines(slingshot.object, lwd = 2, col = 'black')
}


#' Generates a barplot indicating the n best and worst correlated genes by Spearman correlation
#'
#' @param gene Gene for which Spearman correlation will be plotted
#' @param input input dataset, can be a Seurat object or precomputed correlation matrix
#' @param input.type Input type of input datasets. Arguments can be 'correlation.mtx' or 'Seurat'
#' @param number number of best and worst correlated genes to be returned
#' @param color1 Color for best correlated genes
#' @param color2 Color for worst correlated genes
#' @param only.TF only.TF = FALSE plots all genes, only.TF = TRUE plots only TFs (this requires a vector of TFs called TF.lst)
#' @param highlight.gene name of the gene, which is supposed to be highlighted on the plot
#' @param highlight.color color with which the name or the color of highlight.gene are supposed to be indicated
#' @return A ggplot barplot. Bars indicate the Spearman correlation values for the indicated genes with the chosen gene. The chosen gene is automatically removed from the plot. 

plot.corr <- function(gene,
                      input = seurat,
                      input.type = c("correlation.mtx", "Seurat"),
                      number = 20, color1 = "blue",
                      color2 = "orange", only.TF = FALSE,
                      highlight.gene = "Neurod2",
                      highlight.color = "red") {
  if (input.type == "correlation.mtx") {
    dataframe <- input
  } else if (input.type == "Seurat") {
    data.mtx <- as.matrix(GetAssayData(input))

    cat("Calculate correlation ranks!")
    gene.ordered.correlations <- cbind(
      rownames(data.mtx),
      unlist(pblapply(rownames(input), function(genename) {
        return(cor(x = data.mtx[gene, ], y = data.mtx[genename, ], method = "spearman"))
      }))
    )

    colnames(gene.ordered.correlations) <- c("gene", "correlation")

    gene.ordered.correlations <- data.frame(gene.ordered.correlations) %>%
      dplyr::filter(!is.na(correlation)) %>%
      dplyr::mutate(correlation = as.numeric(correlation))

    colnames(gene.ordered.correlations) <- c("genenames", gene)

    rownames(gene.ordered.correlations) <- gene.ordered.correlations$genenames

    dataframe <- t(gene.ordered.correlations)
  } else {
    print("input.type must be defined as Seurat or correlation.mtx")
  }

  selected.row <- as.data.frame(dataframe[gene, ])
  rownames(selected.row) <- make.names(colnames(dataframe), unique = TRUE)
  selected.row[, 2] <- row.names(selected.row)
  selected.row <- selected.row[!is.na(selected.row[, 1]), ]

  if (only.TF == TRUE) {
    selected.row <- selected.row[intersect(rownames(selected.row), TF.lst), ]
  }

  colnames(selected.row) <- c("correlation", "gene_names")
  selected.row$correlation <- as.numeric(selected.row$correlation)

  selected.row <- selected.row %>%
    dplyr::arrange(desc(correlation)) ## order elements

  n <- number + 1
  corr.genes <- rbind(
    selected.row[2:n, ],
    tail(selected.row, n = number)
  )

  if (max(corr.genes$correlation) > 0.6) {
    max.axis <- 1
  } else {
    max.axis <- 0.6
  }

  for (j in 1:nrow(corr.genes)) {
    if (as.numeric(corr.genes[j, "correlation"]) > 0) {
      corr.genes[j, "position"] <- -0.02
      corr.genes[j, "hjust"] <- 1
      corr.genes[j, "color"] <- color1
    }
    else {
      corr.genes[j, "position"] <- 0.02
      corr.genes[j, "hjust"] <- 0
      corr.genes[j, "color"] <- color2
    }
  }

  if (highlight.gene %in% rownames(corr.genes)) {
    corr.genes[highlight.gene, "color"] <- highlight.color
  } else {
    print("highlight.gene not in dataframe !!!")
  }

  textsize <- 240 / number

  bar.plot <- ggplot(data = corr.genes) +
    geom_bar(
      stat = "identity", aes(
        x = reorder(row.names(corr.genes), -corr.genes[, 1]),
        y = corr.genes[, 1],
        fill = corr.genes[, 5]
      ),
      position = "dodge"
    ) +
    scale_fill_manual(values = c(color1, color2, highlight.color)) +
    geom_text(aes(
      x = reorder(row.names(corr.genes), -corr.genes[, 1]),
      y = corr.genes[, 3],
      label = row.names(corr.genes),
      angle = 90,
      hjust = corr.genes[, 4],
      size = 60,
      color = color,
      fontface = "bold"
    )) +
    scale_color_manual(values = c(color1, color2, highlight.color)) +
    ggtitle(gene) +
    theme(
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      axis.title.x = element_blank(),
      axis.text.y = element_text(size = 16),
      axis.title.y = element_text(size = 12),
      legend.title = element_blank(),
      legend.position = "none",
      panel.border = element_blank(),
      panel.background = element_blank(),
      panel.grid.minor.y = element_line(size = 0.1, colour = "black"),
      panel.grid.major.y = element_line(size = 0.2, colour = "black"),
      plot.title = element_text(face = "bold", size = 32)
    ) +
    ylab("Spearman correlation") +
    scale_y_continuous(breaks = pretty_breaks(6), limits = c(-max.axis, max.axis))
  return(bar.plot) ### returns to outputs bar.plot <- ggplot object, gene.list <- correlations of all genes
}

#' Generates a hierarchical clustered correlation matrix
#'
#' @param input input dataset, can be a Seurat object or precomputed correlation matrix
#' @param input.type Input type of input datasets. Arguments can be 'correlation.mtx' or 'Seurat'
#' @param correlation.genes Vector of genes that will be used to compute the correlation matrix 
#' @param min Minimum correlation values to be plotted. Smaller correlations will be shown as grey
#' @param max Maximum correlation values to be plotted. Bigger correlations will be shown as grey
#' @return A ggplot object. The color code indicates Spearman correlation values between the genes

plot.correlation.mtx <- function(input = seurat,
                                 input.type = c("Seurat", "correlation.mtx"),
                                 correlation.genes = c(
                                   "Onecut1", "Onecut2", "Onecut3", "Pou2f2", "Zfhx2", "Zfhx3",
                                   "Zfhx4", "Nfia", "Nfib", "Nfix", "Neurod2", "Neurod6", "Tcf4"
                                 ),
                                 min = -0.6,
                                 max = 0.6) {
  reorder_cormat <- function(cormat) {
    # Use correlation between variables as distance
    dd <- as.dist((1 - cormat) / 2)
    hc <- hclust(dd)
    return(cormat[hc$order, hc$order])
  }

  if (input.type == "correlation.mtx") {
    correlation.mtx <- input[correlation.genes, correlation.genes]
  } else if (input.type == "Seurat") {
    correlation.mtx <- cor(t(as.matrix(GetAssayData(input))[correlation.genes, ]), method = "spearman")
  } else {
    print("input.type must be defined as Seurat or correlation.mtx")
  }

  diag(correlation.mtx) <- NA

  correlation.mtx <- correlation.mtx %>%
    reorder_cormat() %>%
    as.data.frame() %>%
    tibble::rownames_to_column("gene") %>%
    tidyr::gather(key = "gene2", value = "correlation", -gene)

  correlation.mtx$gene <- factor(correlation.mtx$gene, levels = unique(correlation.mtx$gene))
  correlation.mtx$gene2 <- factor(correlation.mtx$gene2, levels = unique(correlation.mtx$gene2))

  correlation.plot <- ggplot(data = correlation.mtx, aes(x = gene, y = gene2, fill = correlation)) +
    geom_tile() +
    scale_fill_gradient2(high = "red", mid = "white", low = "blue", limits = c(min, max), na.value = "grey50") +
    theme_bw() +
    labs(fill = "") +
    theme(
      axis.title = element_blank(),
      axis.text.x = element_text(angle = 90, hjust = 1),
      axis.text = element_text(face = "italic"),
      panel.grid.major = element_blank(),
      legend.position = "top",
      aspect.ratio = 1
    )

  return(correlation.plot)
}

#' Generates a correlation rank plot
#'
#' @param input input dataset, can be a Seurat object or precomputed correlation matrix
#' @param input.type Input type of input datasets. Arguments can be 'correlation.mtx' or 'Seurat'
#' @param plot.gene Gene for which the correlation rank plot will be computed
#' @param correlation.genes Genes that will be highlighted on the correlation rank plot (larger circles and labelled by name) 
#' @param min Minimum correlation values to be plotted. Smaller correlations will be shown as grey
#' @param max Maximum correlation values to be plotted. Bigger correlations will be shown as grey
#' @return A ggplot object indicating Spearman correlation between all genes in the genome and the chosen gene. Data points corresponding to genes passed to correlation.genes will be highlighted and labelled.


plot.correlation.ranks <- function(input = seurat,
                                   input.type = c("Seurat", "correlation.mtx"),
                                   plot.gene = "Zfhx3",
                                   correlation.genes = c(
                                     "ONECUT1", "ONECUT2", "ONECUT3", "POU2F2", "ZFHX2", "ZFHX3",
                                     "ZFHX4", "NFIA", "NFIB", "NFIX", "NEUROD2", "NEUROD6", "TCF4"
                                   ),
                                   min = -0.6,
                                   max = 0.6) {
  if (input.type == "correlation.mtx") {
    gene.ordered.correlations <- data.frame(sort(input[plot.gene, ])) %>%
      tibble::rownames_to_column("gene") %>%
      dplyr::filter(gene != plot.gene)

    colnames(gene.ordered.correlations) <- c("gene", "correlation")
  } else if (input.type == "Seurat") {
    data.mtx <- as.matrix(GetAssayData(input))

    print("Calculate correlation ranks!")
    gene.ordered.correlations <- cbind(
      rownames(data.mtx),
      unlist(pblapply(rownames(input), function(genename) {
        return(cor(x = data.mtx[plot.gene, ], y = data.mtx[genename, ], method = "spearman"))
      }))
    )

    colnames(gene.ordered.correlations) <- c("gene", "correlation")

    gene.ordered.correlations <- data.frame(gene.ordered.correlations) %>%
      dplyr::filter(!is.na(correlation)) %>%
      dplyr::mutate(correlation = as.numeric(correlation)) %>%
      dplyr::filter(gene != plot.gene) %>%
      dplyr::arrange(desc(correlation))
  } else {
    print("input.type must be defined as Seurat or correlation.mtx")
  }

  gene.ordered.correlations$gene <- factor(gene.ordered.correlations$gene, levels = gene.ordered.correlations$gene)

  index <- unlist(lapply(correlation.genes[-which(correlation.genes == plot.gene)], function(x) {
    which(gene.ordered.correlations$gene == x)
  }))

  gene.ordered.correlations$label <- NA
  gene.ordered.correlations$label[index] <- as.character(gene.ordered.correlations$gene[index])

  rank.plot <- ggplot(gene.ordered.correlations, aes(x = gene, y = correlation, label = label)) +
    geom_point(
      color = ifelse(is.na(gene.ordered.correlations$label), "grey70", "red"),
      size = ifelse(is.na(gene.ordered.correlations$label), 1, 3)
    ) +
    geom_text_repel(min.segment.length = 0, seed = 42, max.overlaps = Inf, box.padding = 1, fontface = "italic") +
    scale_x_discrete(expand = c(0.05, 0)) +
    theme_bw() +
    xlab("Gene Rank") +
    ylab("Spearman Correlation") +
    ylim(c(min, max)) +
    ggtitle(plot.gene) +
    theme(
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      panel.grid.major.x = element_blank()
    )

  return(rank.plot)
}


#' Plots gene expression and percentage of expressing cells in certain groups (logic of size and color coding of the dots is inverted relative to the DotPlot function in Seurat)
#'
#' @param input Seurat input dataset
#' @param genes Genes that are supposed to be plotted
#' @param time Seurat Ident by which data will be splitted to calculate mean gene expression and fraction of cells 
#' @return A ggplot object with normalized mean gene expression (indicated by the size of the circles) and percentage of expressing cells (indicated from blue to yellow).

plot.expression.dynamics.from.Seurat <- function(input, genes, time, title) {
  
  num.cells.expressing <- do.call(cbind, lapply(genes, function(x) {
    
    exprs.values <- as.data.frame(GetAssayData(object = input, slot = 'counts'))[x, ] %>%
      t() %>% as.data.frame() %>%
      dplyr::mutate(age = input[[time]]) %>%  
      dplyr::group_by(age)
    
    colnames(exprs.values)[1] <- "expression"
    
    exprs.values <- exprs.values %>%
      dplyr::group_by(age) %>%
      dplyr::summarise(count2 = length(expression[expression > 0]))
    
    return(exprs.values$count2)
  }))
  
  colnames(num.cells.expressing) <- genes
  
  perc.expressing <-  as.data.frame(GetAssayData(object = input, slot = 'counts'))[genes, ] %>%
    t() %>% as.data.frame() %>%
    dplyr::mutate(age = input[[time]]) %>%  
    dplyr::group_by(age) %>%
    dplyr::count() %>%
    dplyr::bind_cols(. ,as_tibble(num.cells.expressing)) %>% as.data.frame() %>%
    dplyr::mutate_if(is.numeric, function(v){v / .[,"n"]}) %>%
    dplyr::select(-n) %>%
    tidyr::gather(key = gene, value = "percent", -age) %>%
    dplyr::pull(percent)
  
  ###normalization function
  normalit<-function(x){
    x/(max(x))
  }
  
  mat <- as.data.frame(GetAssayData(object = input, slot = 'counts'))[genes, ] %>%
    t() %>% as.data.frame() %>%
    dplyr::mutate(age = input[[time]]) %>%
    dplyr::group_by(age) %>%
    dplyr::summarise_all(mean) %>% 
    dplyr::mutate_each(funs(normalit), all_of(genes)) %>%
    as.data.frame() %>%
    tidyr::gather(key = gene, value = "readcounts", -age) %>%
    dplyr::mutate(percent = perc.expressing)
  
  mat$gene <- factor(mat$gene, levels = genes)
  
  gg <- ggplot(data = as.data.frame(lapply(mat, unlist)), aes(x = factor(age), y = factor(gene))) +
    geom_count(aes(size = readcounts, color = percent)) +
    scale_size_area(max_size = 10) +
    scale_color_gradientn(colours = c("blue", "yellow"), limits = c(0,1)) +
    theme_classic() +
    ggtitle(title) +
    theme(axis.text.y = element_text(size = 24, face = 'italic'),
          axis.title = element_blank(),
          axis.text.x = element_text(size = 24, angle = 45, hjust = 1, vjust = 1),
          legend.title = element_text(size = 20),
          legend.title.align=0.5,
          legend.text = element_text(size = 20),
          strip.text = element_text(size = 24, face = "bold"),
          plot.title = element_text(face = "bold", size = 24))  
  
  return(gg)
}

```

## Connect loom file from La Manno et al. 2020. 

```{r Load loom file, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}
## connect sc.loom file downloaded from mousebrain.org
sc.loom <- loomR::connect(filename = paste0(dir, "/input/dev_all.loom"), mode = 'r+', skip.validate = TRUE)

## Generate sc.meta file by extracting parameters from connected sc.loom file
sc.meta <- data.frame(sc.loom$col.attrs$Age[], 
                      sc.loom$col.attrs$PseudoAge[], 
                      sc.loom$col.attrs$Tissue[], 
                      sc.loom$col.attrs$PseudoTissue[],
                      sc.loom$col.attrs$Class[],
                      sc.loom$col.attrs$Clusters[],
                      10000 / sc.loom$col.attrs$TotalUMI[],
                      sc.loom$col.attrs$CellID[],
                      sc.loom$col.attrs$SampleID[])

colnames(sc.meta) <- c("age", "pseudoage", "tissue", "pseudotissue", "class", "clusters", "normalization_factor", "cellID", 'sampleID')
```


## Late middbrain stages / comparison tTFs

```{r Load midbrain data late neurons, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}

tissue = "Midbrain" 
celltype = "Neuron"
timepoints = c('e16.0', 'e16.25', 'e16.5', 'e17.0', 'e17.5', 'e18.0')

tissue.id <- which(grepl(tissue, unique(sc.loom$col.attrs$Tissue[])) == TRUE)
cell.id <- intersect(which(sc.meta$tissue %in% unique(sc.loom$col.attrs$Tissue[])[tissue.id] & sc.meta$class == celltype),
                     which(sc.meta$age %in% timepoints))

exp.mat <- sc.loom[["matrix"]][cell.id,]

colnames(exp.mat) <- sc.loom$row.attrs$Gene[]
rownames(exp.mat) <- sc.meta$cellID[cell.id]

mb.seurat.late <- CreateSeuratObject(counts = t(exp.mat), 
                                meta.data = sc.meta[cell.id,] %>% 
                                  as.tibble() %>% 
                                  tibble::column_to_rownames("cellID"))

mb.seurat.late[["percent.mt"]] <- PercentageFeatureSet(mb.seurat.late, pattern = "^mt-")

mb.seurat.late <- mb.seurat.late %>% 
  subset(subset = nFeature_RNA > 600 & nFeature_RNA < 6000 & percent.mt < 6) %>%
  SCTransform(vars.to.regress = 'sampleID') %>%
  NormalizeData(verbose=FALSE) %>%
  ScaleData(verbose = FALSE) %>%
  FindVariableFeatures(selection.method = "vst", verbose = FALSE) %>%
  RunPCA(npcs = 30, verbose = FALSE) %>%
  RunUMAP(reduction = "pca", dims = 1:30) %>%
  FindNeighbors(dims = 1:30) %>%
  FindClusters(resolution = 0.5)
```

```{r UMAP of late midbrain neurons, echo = TRUE, fig.width = 15, fig.height = 4}

mb.umap.late <- cowplot::plot_grid(DimPlot(mb.seurat.late, reduction = 'umap', group.by = "age") + ggtitle("Age"),
                   FeaturePlot(mb.seurat.late, features = c('Pou2f2', 'Zfhx3', 'Zfhx4', 'Tcf4', 
                                                            'Nfia', 'Nfib', 'Neurod2', 'Neurod6'), 
                               ncol = 4) & NoLegend() & NoAxes() & theme(plot.title = element_text(face = 'italic')),
                   nrow = 1, rel_widths = c(1,2))

mb.umap.late
```

```{r Correlation plot matrix amacrine / horizontal cells, echo=TRUE, message=FALSE, warning=FALSE, results='hide', linewidth = 90, fig.width=3, fig.height=3, fig.keep='all'}

mb.corr.mtx <- plot.correlation.mtx(
  input = mb.seurat.late,
  input.type = c("Seurat"),
  correlation.genes = c(
    "Pou2f2", "Zfhx3", "Zfhx4", "Nfia", "Nfib", "Nfix", "Neurod2", 'Neurod6', "Tcf4"
  ),
  min = -0.8,
  max = 0.8
)

mb.corr.mtx

```


```{r Correlation rank plots late midbrain cells, echo=TRUE, message=FALSE, warning=FALSE, results='hide', linewidth = 90, fig.width=3, fig.height=6, fig.keep='all'}

mb.corr.rank <- cowplot::plot_grid(
  plot.correlation.ranks(
    input = mb.seurat.late,
    input.type = c("Seurat"),
    plot.gene = 'Zfhx3',
    correlation.genes = c(
      "Pou2f2", "Zfhx3", "Zfhx4", "Nfia", "Nfib", "Nfix", "Neurod2", "Neurod6", "Tcf4"
    ),
    min = -0.6,
    max = 0.6
  ),
  plot.correlation.ranks(
    input = mb.seurat.late,
    input.type = c("Seurat"),
    plot.gene = 'Nfib',
    correlation.genes = c(
      "Pou2f2", "Zfhx3", "Zfhx4", "Nfia", "Nfib", "Nfix", "Neurod2", "Neurod6", "Tcf4"
    ),
    min = -0.8,
    max = 0.8
  ),
  ncol = 1
)

mb.corr.rank
```

```{r Combine late midbrain plots, echo=TRUE, message=FALSE, warning=FALSE, results='hide', linewidth = 90, fig.width=15, fig.height=8, fig.keep='all'}

bottom <- cowplot::plot_grid(FeaturePlot(mb.seurat.late, features = c('Tubb3', 'Elavl3', 'S100b', 'Slc1a3')) & 
                               NoLegend() & NoAxes() & theme(plot.title = element_text(face = 'italic')),
                             mb.corr.mtx, 
                             mb.corr.rank + theme(aspect.ratio = 1) & theme(plot.title = element_text(face = 'italic')), 
                             nrow = 1)

mb.plot <- cowplot::plot_grid(mb.umap.late, bottom,
                    nrow = 2)

mb.plot
```

## Late forebrain stages / comparison tTFs

```{r Load forebrain data late neurons, echo=TRUE, message=FALSE, warning=FALSE, results='hide'}

tissue = "Forebrain" 
celltype = "Neuron"
timepoints = c('e16.0', 'e16.25', 'e16.5', 'e17.0', 'e17.5', 'e18.0')

tissue.id <- which(grepl(tissue, unique(sc.loom$col.attrs$Tissue[])) == TRUE)
cell.id <- intersect(which(sc.meta$tissue %in% unique(sc.loom$col.attrs$Tissue[])[tissue.id] & sc.meta$class == celltype),
                     which(sc.meta$age %in% timepoints))

exp.mat <- sc.loom[["matrix"]][cell.id,]

colnames(exp.mat) <- sc.loom$row.attrs$Gene[]
rownames(exp.mat) <- sc.meta$cellID[cell.id]

fb.seurat.late <- CreateSeuratObject(counts = t(exp.mat), 
                                meta.data = sc.meta[cell.id,] %>% 
                                  as.tibble() %>% 
                                  tibble::column_to_rownames("cellID"))

fb.seurat.late[["percent.mt"]] <- PercentageFeatureSet(fb.seurat.late, pattern = "^mt-")

fb.seurat.late <- fb.seurat.late %>% 
  subset(subset = nFeature_RNA > 600 & nFeature_RNA < 6000 & percent.mt < 6) %>%
  SCTransform(vars.to.regress = 'sampleID') %>%
  NormalizeData(verbose=FALSE) %>%
  ScaleData(verbose = FALSE) %>%
  FindVariableFeatures(selection.method = "vst", verbose = FALSE) %>%
  RunPCA(npcs = 30, verbose = FALSE) %>%
  RunUMAP(reduction = "pca", dims = 1:30) %>%
  FindNeighbors(dims = 1:30) %>%
  FindClusters(resolution = 0.5)
```

```{r UMAP of late forebrain neurons, echo = TRUE, fig.width = 15, fig.height = 4}

fb.umap.late <- cowplot::plot_grid(DimPlot(fb.seurat.late, reduction = 'umap', group.by = "age") + ggtitle("Age"),
                   FeaturePlot(fb.seurat.late, features = c('Pou2f2', 'Zfhx3', 'Zfhx4', 'Tcf4', 
                                                            'Nfia', 'Nfib', 'Neurod2', 'Neurod6'), 
                               ncol = 4) & NoLegend() & NoAxes() & theme(plot.title = element_text(face = 'italic')),
                   nrow = 1, rel_widths = c(1,2))

fb.umap.late
```

```{r Correlation plot matrix late forebrain neurons, echo=TRUE, message=FALSE, warning=FALSE, results='hide', linewidth = 90, fig.width=3, fig.height=5, fig.keep='all'}

fb.corr.mtx <- plot.correlation.mtx(
  input = fb.seurat.late,
  input.type = c("Seurat"),
  correlation.genes = c(
    "Pou2f2", "Zfhx3", "Zfhx4", "Nfia", "Nfib", "Nfix", "Neurod2", 'Neurod6', "Tcf4"
  ),
  min = -0.8,
  max = 0.8
)

fb.corr.mtx

```

```{r Correlation rank plots late forebrain cells, echo=TRUE, message=FALSE, warning=FALSE, results='hide', linewidth = 90, fig.width=3, fig.height=6, fig.keep='all'}

fb.corr.rank <- cowplot::plot_grid(
  plot.correlation.ranks(
    input = fb.seurat.late,
    input.type = c("Seurat"),
    plot.gene = 'Zfhx3',
    correlation.genes = c(
      "Pou2f2", "Zfhx3", "Zfhx4", "Nfia", "Nfib", "Nfix", "Neurod2", "Neurod6", "Tcf4"
    ),
    min = -0.6,
    max = 0.6
  ),
  plot.correlation.ranks(
    input = fb.seurat.late,
    input.type = c("Seurat"),
    plot.gene = 'Nfib',
    correlation.genes = c(
      "Pou2f2", "Zfhx3", "Zfhx4", "Nfia", "Nfib", "Nfix", "Neurod2", "Neurod6", "Tcf4"
    ),
    min = -0.8,
    max = 0.8
  ),
  nrow = 2
)

fb.corr.rank
```

```{r Combine late forebrain plots, echo=TRUE, message=FALSE, warning=FALSE, results='hide', linewidth = 90, fig.width=15, fig.height=8, fig.keep='all'}

bottom <- cowplot::plot_grid(FeaturePlot(fb.seurat.late, features = c('Tubb3', 'Elavl3', 'S100b', 'Slc1a3')) & 
                               NoLegend() & NoAxes() & theme(plot.title = element_text(face = 'italic')),
                             fb.corr.mtx, 
                             fb.corr.rank + theme(aspect.ratio = 1) & theme(plot.title = element_text(face = 'italic')), 
                             nrow = 1)

fb.plot <- cowplot::plot_grid(fb.umap.late, bottom,
                    nrow = 2)

fb.plot
```

```{r Combine late plots, echo=TRUE, message=FALSE, warning=FALSE, results='hide', linewidth = 90, fig.width=15, fig.height=16, fig.keep='all'}

cowplot::plot_grid(mb.plot, fb.plot, nrow = 2)

```

```{r Include sessionInfo, echo = TRUE, fig.width = 10, fig.height = 6}

sessionInfo()

```